pkg/boot/uefi/uefi.go

```diff
@@ -0,0 +1,180 @@
+	name         string
+	mem          kexec.Memory
+	entryAddress uintptr
+	ImageBase    uintptr
+	SerialConfig SerialPortConfig
+}
+
+func checkFvAndGetEntryPoint(name string) (uintptr, error) {
+	f, err := os.Open(name)
+	if err != nil {
+		return 0, err
+	}
+	defer f.Close()
+	// Entry PE image for UEFI Payload is 0xA0 offset from FV head.
+	pef := io.NewSectionReader(f, fvEntryImageOffset, math.MaxInt64)
+	pf, err := pe.NewFile(pef)
+	if err != nil {
+		return 0, err
+	}
+	op64, ok := pf.OptionalHeader.(*pe.OptionalHeader64)
+	if !ok {
+		return 0, fmt.Errorf("It is not OptionalHeader64")
+	}
+	return uintptr(fvEntryImageOffset) + uintptr(op64.AddressOfEntryPoint), nil
+}
+
+// New loads the file and return FvImage stucture if entry image is found
+func New(n string) (*FvImage, error) {
+	entry, err := checkFvAndGetEntryPoint(n)
+	if err != nil {
+		return nil, err
+	}
+	return &FvImage{name: n, mem: kexec.Memory{}, entryAddress: entry}, nil
+}
+
+// Reserved 64kb for passing params
+const uefiPayloadConfigSize = 0x10000
+
+// Load loads fimware volume payload and boot the the payload
+func (fv *FvImage) Load(verbose bool) error {
+	// Install payload
+	dat, err := ioutil.ReadFile(fv.name)
+	if err != nil {
+		return err
+	}
+	fv.mem.Segments.Insert(kexec.NewSegment(dat, kexec.Range{Start: fv.ImageBase, Size: uint(len(dat))}))
+
+	// Install payload config & its memory map: 64 kb below the image
+	// We cannot use the memory above the image base because it may be used by HOB
+	var configAddr uintptr = fv.ImageBase - uintptr(uefiPayloadConfigSize)
+
+	// Get MemoryMap
+	mm, err := kexecParseMemoryMap()
+	if err != nil {
+		return err
+	}
+
+	// Get Acpi Basc (RSDP)
+	rsdp, err := getRSDP()
+	if err != nil {
+		return err
+	}
+
+	smbiosBase, smbiosSize, err := getSMBIOSBase()
+	if err != nil {
+		return err
+	}
+
+	pc := payloadConfig{
+		Version:             PayloadConfigVersion,
+		AcpiBase:            uint64(rsdp.RSDPAddr()),
+		AcpiSize:            uint64(rsdp.Len()),
+		SmbiosBase:          uint64(smbiosBase),
+		SmbiosSize:          uint64(smbiosSize),
+		SerialConfig:        fv.SerialConfig,
+		NumMemoryMapEntries: uint32(len(mm)),
+	}
+
+	pcbuf := &bytes.Buffer{}
+	if err := binary.Write(pcbuf, binary.LittleEndian, pc); err != nil {
+		return err
+	}
+
+	if err := binary.Write(pcbuf, binary.LittleEndian, mm.AsPayloadParam()); err != nil {
+		return err
+	}
+
+	if len(pcbuf.Bytes()) > uefiPayloadConfigSize {
+		return fmt.Errorf("Config/Memmap size is greater than reserved size: %d bytes", len(pcbuf.Bytes()))
+	}
+
+	fv.mem.Segments.Insert(kexec.NewSegment(pcbuf.Bytes(), kexec.Range{Start: configAddr, Size: uint(len(pcbuf.Bytes()))}))
+
+	if verbose {
+		log.Printf("segments cmdline %v", fv.mem.Segments)
+	}
+
+	if err := kexecLoad(fv.ImageBase+uintptr(fv.entryAddress), fv.mem.Segments, 0); err != nil {
+		return fmt.Errorf("kexec.Load() error: %v", err)
+	}
+
+	if err := bootExecute(); err != nil {
+		log.Fatal(err)
+	}
```

Comment:
leave the bootExecute call to the caller. Load() should only load -- that's what the multiboot loader does. then you can abolish the bootExecute dependency injection too.

Reply:
Done.